### PR TITLE
[RFC] Nested Templates (and wrapping JavaScript with templates)

### PR DESCRIPTION
**Pull request by jbolda**
This is a PR based on Issue #1866. Effectively updating createPages to accept an array of templates (and a javascript Post type of page). I got as much done as I could, and I probably won't be able to work on it for a while. Hopefully someone else can pick up where I left off. The using-javascript-transforms should be updated to work on this PR. I have been using it to test.

The remaining TODO as I am aware:
- [x] accept array for createPages component in develop
- [x] fix the props (and specifically props.data from graphql) as it gets all munged between the templates
- [x] `gatsby build`, ie have it run in production
- [x] update old tests
  - [x] gatsby-transformer-documentationjs seems to [randomly fail](https://ci.appveyor.com/project/KyleAMathews/gatsby/build/1.0.2697)
- [x] add tests for the new situations
- [x] dev 404 appears broken, fix
- [x] update using-javascript-transforms example templates to pull out blog post structure into overall component
- [ ] update documentation

### CODE REVIEW

**Comment by jbolda** [comment]
The most recent commits seem to fix the data issue. Some more testing then, then to look at production code.

File: examples/using-javascript-transforms/src/articles/2017-03-09-choropleth-on-d3v4/index.js [hunk 1]
```diff
@@ -47,6 +47,7 @@ class choroplethBase extends React.Component {
   }
 
   render() {
+    console.log(this)
```
Comment (nihgwu) [inline examples/using-javascript-transforms/src/articles/2017-03-09-choropleth-on-d3v4/index.js]: why `console.log` here

File: examples/using-javascript-transforms/src/articles/2017-05-30-choropleth-on-d3v4-alternate/index.js [hunk 2]
```diff
@@ -46,6 +46,7 @@ class choroplethAltBase extends React.Component {
   }
 
   render() {
+    console.log(this)
```
Comment (nihgwu) [inline examples/using-javascript-transforms/src/articles/2017-05-30-choropleth-on-d3v4-alternate/index.js]: why `console.log` here

File: examples/using-javascript-transforms/src/articles/2017-03-09-choropleth-on-d3v4/index.js (hunk 1 above)
Comment (jbolda) [reply examples/using-javascript-transforms/src/articles/2017-03-09-choropleth-on-d3v4/index.js]: Using them to test, and forgot to delete them.

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit 36a280dc83d89017e4da35f2a374abe760fb790a

https://app.netlify.com/sites/using-glamor/deploys/59b1e8960752d07ffbc40597

**Comment by jbolda** [comment]
Nearing the end of the tunnel here. The technical implementation for the template takes an array or coerces a string into an array. When creating the elements, we recursively loop through the array to best the components. The data is still per page, but it is now separated by keys based on the componentChunkName. The data is stored this way, but retrieved in a manner that returns the same data structure end result. The idea is for this to be compatible with any existing implementations.

I am keeping the TODO in the OP updated as I go.

**Comment by KyleAMathews** [comment]
Hey just had a quick look at this and wanted to correct a miscommunication. We want nested *layout* components not nested *page* components. It's perhaps more semantics than anything else but there can only be one page component controlling the page at a time. The boundary that's significant is between pages and layouts as layouts stay constant across page changes which lets you retain state in the layout as you click between pages. Having nested page components doesn't gain you anything because you can already nest components quite easily yourself.

**Comment by KyleAMathews** [comment]
On train writing on my phone so happy to discuss at more length later :-)

**Comment by jbolda** [comment]
This is substantially complete sans documentation. The discussion is in #1866 (and follow-ups to @KyleAMathews comment above). Comments, code reviews, and testing on your own website are welcome. The idea is that this is effectively backward compatible to existing websites (no breaking changes). I can update documentation when the current implementation is greenlit.

**Comment by jbolda** [comment]
Closed per discussion on #1866.

**Comment by nihgwu** [comment]
Can we make it a plugin instead of putting in core? I really  appreciate your efforts on this great feature

**Pull request closed without merging**